Version from 42.1.0 can't connect to 127.0.0.1 with System SOCKS proxy

Hi,
I found Postgresql JDBC Driver cannot connect to local postgresql server when I set a system SOCKS proxy.

org.postgresql.util.PSQLException: The connection attempt failed.
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:275)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:194)
	at org.postgresql.Driver.makeConnection(Driver.java:450)
	at org.postgresql.Driver.connect(Driver.java:252)
	at java.sql.DriverManager.getConnection(DriverManager.java:664)
	at java.sql.DriverManager.getConnection(DriverManager.java:208)
	at org.apache.ibatis.datasource.unpooled.UnpooledDataSource.doGetConnection(UnpooledDataSource.java:201)
	at org.apache.ibatis.datasource.unpooled.UnpooledDataSource.doGetConnection(UnpooledDataSource.java:196)
	at org.apache.ibatis.datasource.unpooled.UnpooledDataSource.getConnection(UnpooledDataSource.java:93)
	at org.apache.ibatis.datasource.pooled.PooledDataSource.popConnection(PooledDataSource.java:404)
.......

Caused by: java.net.UnknownHostException: 127.0.0.1
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:184)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.postgresql.core.PGStream.<init>(PGStream.java:68)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:144)
	... 69 more

system:  macOS High Sierra
java: 1.8.0_144

seeing this same behavior with version `42.1.4`.
```
[HikariPool-1 connection adder] ERROR org.postgresql.ds.common.BaseDataSource - Failed to create a Non-Pooling DataSource from PostgreSQL JDBC Driver 42.1.4 for fwptools at jdbc:postgresql://172.16.3.128:32774/fwptools?loginTimeout=30: org.postgresql.util.PSQLException: The connection attempt failed.
[HikariPool-1 connection adder] ERROR org.postgresql.Driver - Connection error: 
org.postgresql.util.PSQLException: The connection attempt failed.
	at org.postgresql.Driver$ConnectThread.getResult(Driver.java:401)
	at org.postgresql.Driver.connect(Driver.java:259)
	at java.sql.DriverManager.getConnection(DriverManager.java:664)
	at java.sql.DriverManager.getConnection(DriverManager.java:247)
	at org.postgresql.ds.common.BaseDataSource.getConnection(BaseDataSource.java:94)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:375)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:204)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:445)
	at com.zaxxer.hikari.pool.HikariPool.access$200(HikariPool.java:72)
	at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:632)
	at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:618)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.UnknownHostException: 172.16.3.128
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:184)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.postgresql.core.PGStream.<init>(PGStream.java:68)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:144)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:194)
	at org.postgresql.Driver.makeConnection(Driver.java:450)
	at org.postgresql.Driver.access$100(Driver.java:60)
	at org.postgresql.Driver$ConnectThread.run(Driver.java:360)
	... 1 more
```
in this case `172.16.3.128` is the ip of my docker vm that the jvm is told to ignore via
```
-DsocksProxyHost=127.0.0.1 -DsocksProxyPort=9998 -DsocksNonProxyHosts=172.16.3.128
```
the driver version that works w/ this configuration is `42.0.0`

looking through the changelog history, i see a reference to #774 which i am guessing is the cause of this behavior.

--- a/pgjdbc/src/main/java/org/postgresql/core/PGStream.java
+++ b/pgjdbc/src/main/java/org/postgresql/core/PGStream.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2003, PostgreSQL Global Development Group
+ * Copyright (c) 2017, PostgreSQL Global Development Group
  * See the LICENSE file in the project root for more information.
  */
 
@@ -21,7 +21,6 @@ import java.io.Writer;
 import java.net.InetSocketAddress;
 import java.net.Socket;
 import java.sql.SQLException;
-
 import javax.net.SocketFactory;
 
 /**
@@ -63,7 +62,7 @@ public class PGStream implements Closeable {
       // When using a SOCKS proxy, the host might not be resolvable locally,
       // thus we defer resolution until the traffic reaches the proxy. If there
       // is no proxy, we must resolve the host to an IP to connect the socket.
-      InetSocketAddress address = System.getProperty("socksProxyHost") == null
+      InetSocketAddress address = hostSpec.shouldResolve()
           ? new InetSocketAddress(hostSpec.getHost(), hostSpec.getPort())
           : InetSocketAddress.createUnresolved(hostSpec.getHost(), hostSpec.getPort());
       socket.connect(address, timeout);

--- a/pgjdbc/src/main/java/org/postgresql/util/HostSpec.java
+++ b/pgjdbc/src/main/java/org/postgresql/util/HostSpec.java
@@ -5,10 +5,17 @@
 
 package org.postgresql.util;
 
+import static java.util.regex.Pattern.compile;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 /**
  * Simple container for host and port.
  */
 public class HostSpec {
+  public static String DEFAULT_NON_PROXY_HOSTS = "localhost|127.*|[::1]|0.0.0.0|[::0]";
+
   protected final String host;
   protected final int port;
 
@@ -38,5 +45,51 @@ public class HostSpec {
   @Override
   public int hashCode() {
     return port ^ host.hashCode();
+  }
+
+  public Boolean shouldResolve() {
+    if (System.getProperty("socksProxyHost") == null) {
+      return true;
+    }
+    return matchesNonProxyHosts();
+  }
+
+  private Boolean matchesNonProxyHosts() {
+    String nonProxyHosts = System.getProperty("socksNonProxyHosts", DEFAULT_NON_PROXY_HOSTS);
+    if (nonProxyHosts == null || this.host.isEmpty()) {
+      return false;
+    }
+
+    Pattern pattern = toPattern(nonProxyHosts);
+    Matcher matcher = pattern == null ? null : pattern.matcher(this.host);
+    return matcher != null && matcher.matches();
+  }
+
+  private Pattern toPattern(String mask) {
+    StringBuilder joiner = new StringBuilder();
+    String separator = "";
+    for (String disjunct : mask.split("\\|")) {
+      if (!disjunct.isEmpty()) {
+        String regex = disjunctToRegex(disjunct.toLowerCase());
+        joiner.append(separator).append(regex);
+        separator = "|";
+      }
+    }
+
+    return joiner.length() == 0 ? null : compile(joiner.toString());
+  }
+
+  private String disjunctToRegex(String disjunct) {
+    String regex;
+
+    if (disjunct.startsWith("*")) {
+      regex = ".*" + Pattern.quote(disjunct.substring(1));
+    } else if (disjunct.endsWith("*")) {
+      regex = Pattern.quote(disjunct.substring(0, disjunct.length() - 1)) + ".*";
+    } else {
+      regex = Pattern.quote(disjunct);
+    }
+
+    return regex;
   }
 }

--- a/pgjdbc/src/test/java/org/postgresql/test/jre8/core/Jre8TestSuite.java
+++ b/pgjdbc/src/test/java/org/postgresql/test/jre8/core/Jre8TestSuite.java
@@ -1,0 +1,18 @@
+/*
+ * Copyright (c) 2017, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.test.jre8.core;
+
+import org.junit.runner.RunWith;
+import org.junit.runners.Suite;
+
+/**
+ * @author Joe Kutner on 10/24/17.
+ *         Twitter: @codefinger
+ */
+@RunWith(Suite.class)
+@Suite.SuiteClasses({SocksProxyTest.class})
+public class Jre8TestSuite {
+}

--- a/pgjdbc/src/test/java/org/postgresql/test/jre8/core/SocksProxyTest.java
+++ b/pgjdbc/src/test/java/org/postgresql/test/jre8/core/SocksProxyTest.java
@@ -1,0 +1,48 @@
+/*
+ * Copyright (c) 2017, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.test.jre8.core;
+
+import static org.junit.Assert.assertNotNull;
+
+import org.postgresql.test.TestUtil;
+
+import org.junit.After;
+import org.junit.Test;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+
+/**
+ * @author Joe Kutner on 10/9/17.
+ *         Twitter: @codefinger
+ */
+public class SocksProxyTest {
+
+  @After
+  public void cleanup() {
+    System.clearProperty("socksProxyHost");
+    System.clearProperty("socksProxyPort");
+    System.clearProperty("socksNonProxyHosts");
+  }
+
+  /**
+   * Tests the connect method by connecting to the test database
+   */
+  @Test
+  public void testConnectWithSocksNonProxyHost() throws Exception {
+    System.setProperty("socksProxyHost", "fake-socks-proxy");
+    System.setProperty("socksProxyPort", "9999");
+    System.setProperty("socksNonProxyHosts", TestUtil.getServer());
+
+    TestUtil.initDriver(); // Set up log levels, etc.
+
+    Connection con =
+        DriverManager.getConnection(TestUtil.getURL(), TestUtil.getUser(), TestUtil.getPassword());
+
+    assertNotNull(con);
+    con.close();
+  }
+}

--- a/pgjdbc/src/test/java/org/postgresql/test/util/HostSpecTest.java
+++ b/pgjdbc/src/test/java/org/postgresql/test/util/HostSpecTest.java
@@ -1,0 +1,72 @@
+/*
+ * Copyright (c) 2017, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.test.util;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import org.postgresql.util.HostSpec;
+
+import org.junit.After;
+import org.junit.Test;
+
+/**
+ * @author Joe Kutner on 10/19/17.
+ *         Twitter: @codefinger
+ */
+public class HostSpecTest {
+
+  @After
+  public void cleanup() {
+    System.clearProperty("socksProxyHost");
+    System.clearProperty("socksProxyPort");
+    System.clearProperty("socksNonProxyHosts");
+  }
+
+  @Test
+  public void testShouldResolve() throws Exception {
+    HostSpec hostSpec = new HostSpec("localhost", 5432);
+    assertTrue(hostSpec.shouldResolve());
+  }
+
+  @Test
+  public void testShouldResolveWithSocksProxyHost() throws Exception {
+    System.setProperty("socksProxyHost", "fake-socks-proxy");
+    HostSpec hostSpec = new HostSpec("example.com", 5432);
+    assertFalse(hostSpec.shouldResolve());
+  }
+
+  @Test
+  public void testShouldResolveWithSocksProxyHostWithLocalhost() throws Exception {
+    System.setProperty("socksProxyHost", "fake-socks-proxy");
+    HostSpec hostSpec = new HostSpec("localhost", 5432);
+    assertTrue(hostSpec.shouldResolve());
+  }
+
+  @Test
+  public void testShouldResolveWithSocksNonProxyHost() throws Exception {
+    System.setProperty("socksProxyHost", "fake-socks-proxy");
+    System.setProperty("socksNonProxyHosts", "example.com");
+    HostSpec hostSpec = new HostSpec("example.com", 5432);
+    assertTrue(hostSpec.shouldResolve());
+  }
+
+  @Test
+  public void testShouldResolveWithSocksNonProxyHosts() throws Exception {
+    System.setProperty("socksProxyHost", "fake-socks-proxy");
+    System.setProperty("socksNonProxyHosts", "example.com|localhost");
+    HostSpec hostSpec = new HostSpec("example.com", 5432);
+    assertTrue(hostSpec.shouldResolve());
+  }
+
+  @Test
+  public void testShouldResolveWithSocksNonProxyHostsNotMatching() throws Exception {
+    System.setProperty("socksProxyHost", "fake-socks-proxy");
+    System.setProperty("socksNonProxyHosts", "example.com|localhost");
+    HostSpec hostSpec = new HostSpec("example.org", 5432);
+    assertFalse(hostSpec.shouldResolve());
+  }
+}

